### ISSUE TITLE
testing and improvements for lightgun implementation

### ISSUE THREAD

**Issue by markwkidd** [libretro/mame2003-plus-libretro#1036]
After some delightful refactoring and more importantly some several very useful guidance from mahoneyt944, I am glad to say that there is another selectable input mode for libretro frontends -- "Lightgun + Keyboard Encoder"

This new issue is to track bugs and possible improvements moving forward.

### Wishlist

* **"All-in-One" input mode with lightgun support**
There is at least one thing I have on the wishlist, which is to add the ability in the MAME remapper to simultaneously support input from gamepad, mouse, lightgun, and keyboard. That would allow a "all-at-once" libretro input mode, something not originally available in MAME 0.78.

**Comment by mahoneyt944** [comment]
Is this suppose to use a COMPARE function too or no?

**Comment by markwkidd** [comment]
On the MAME side, there are no dedicated lightgun input codes for the lightgun buttons. Whereas I wish I had something like `JOYCODE_LIGHTGUN_1_BUTTON1` in `input.h` I am currently comparing for `JOYCODE_MOUSE_1_BUTTON1` just like the mouse.

In other words, I think I have the compare macro I need for the current functionality.

I do want to add `JOYCODE_LIGHTGUN_1_BUTTON1` , etc. I know I would need to implement a `seq_set_6` and a `set_seq_7` and look for places where the current limit of 5 sequences could be hardcoded. Ugh! 😄  👯

**Comment by thatman84** [comment]
Hi @markwkidd just duplicating a post here from libretro bounties discord.
Glad your original bounty has now closed out. Hopefully I will get around to testing in the next few weeks. Real life is getting in the way

Anyway... will the recent closure of this mame2003 lightgun bounty have anything transferable more globally relating to the mouse index changing? With the increase in Sinden lightguns users, everyday it would be great to push the whole solution forward.

Ref issue #7638

**Comment by mahoneyt944** [comment]
@markwkidd @thatman84 
I've experimented with mouse indexing using different combinations of input devices. From what what I can tell, the index issue is related to how the input device is connected and how retroarch interprets it.

For example, if I connected a ipac2 with a ultimarc spinner directly to a rpi3, I get mouse index 0. If I plug a USB GRS spinner directly to the rpi3 while still using the ipac2 (without the ultimarc spinner), The mouse index is 2. Seems like keyboard encoders are confusing the indexes in RA maybe?

**Comment by mahoneyt944** [comment]
@markwkidd next issue! Lol apparently the mapping is taking the macro quite literally.
![Screenshot_20210528-103821](https://user-images.githubusercontent.com/49591133/120000790-2ff29900-bfa1-11eb-9577-c5dbb76dbac4.png)

**Comment by markwkidd** [comment]
Labels for P4-P8 fixed.

**Comment by mahoneyt944** [comment]
I'll test more after it builds.

**Comment by mahoneyt944** [comment]
Looks good. Just need to peel back the onion now to clean up anything we don't need with the new plan. Such as this since it's no longer a device type.

https://github.com/libretro/mame2003-plus-libretro/blob/b547860543984b7d43eb95627a86f4a3bf98ad93/src/mame2003/mame2003.c#L1718

**Comment by markwkidd** [comment]
Thank you for the catch. Should be fixed now.

**Comment by mahoneyt944** [comment]
@grant2258

hey, I did some testing using revolution x with swapping the xy_device and the issue is with the pointer device it seems. You can switch between lightgun, mouse and off and also override with analog or digital etc without any scaling issues. 

What I notice with the pointer device is that when your not touching the screen, the crosshair position seems to be up and to the right. I'm speculating that when the pointer is not pressed it's getting an odd value that's breaking the scaling.

**Comment by mahoneyt944** [comment]
so I tested commenting out relative device reads and the pointer still has this issue described above. Need to see if this is an issue with the core or from the pointer input we receive from the frontend.

**Comment by grant2258** [comment]
the abs issues on the mouse have been fixed if your using a pointer device the game self calibrates as well so if the min max abs is different there will be issues

**Comment by mahoneyt944** [comment]
For me on Android everything seems to be working beside the pointer. I can switch between all the others and there are no issues even when overriding the controls and such. Maybe this will be fixed in the issue you linked, idk.

Other than that I haven't changed anything mouse related besides adding the swap XY dial option. I followed up on the request and they where wiring two spinners as one mouse, X and Y. So people requested to use the right sided spinner to work as player 1 X instead of Y. So that's what that was about.

**Comment by grant2258** [comment]
Make sure the device is s dial before swapping the axis the don't translate to a mouse welll. A dial translates better with a keypress

**Comment by mahoneyt944** [comment]
The option is only active if a dial is used. The content flag is used for this. Otherwise the normal code path is used

**Comment by grant2258** [comment]
what exactly is the case use to swap the axis do you have and example ?

**Comment by mahoneyt944** [comment]
Just discussing the "testings" part is all. Things work perfectly fine switching between lightgun and mouse device on the fly. It's the pointer device that messes up the scaling that we talked about before.

The swap XY option is only used for relative devices when a dial is being used. It's 100% optional. This is how the user explained the setup
```
The two spinners on the cab actually belong to a single physical device: Left spinner is the X axis, and right spinner is the Y axis. Together they make a single "mouse" input (and share the same mouse index).

So here in order to use the right spinner for player 1, we need to read the Y axis from the mouse and treat it as the X axis input. The option you added did this exactly. So thanks again for your help!
```

**Comment by grant2258** [comment]
What platform is it happening on. I'll test the game when I get time. I have to get this subframe finished up as well. Have it undercoated just waiting on a new bush before I paint it.

**Comment by mahoneyt944** [comment]
The pointer issue was tested on Android with touch screen.  When you touch it, it tracks correctly. When you are not touching it, it jumps to an arbitrary position and after that point the scaling will be messed up. Maybe this is an input driver issue when the input is "out of bounds".

**Comment by grant2258** [comment]
sounds like something the frontend would be causing youll need to debug what its sending back and also check the calibration values of each device to get an idea of whats going on.

**Comment by grant2258** [comment]
No pointer jumping or scaling issues on udev with my touchpad. Since your talking analog scaling shouldnt be an issue anyway.

here is the rules for the pointer device according to the api

https://github.com/libretro/libretro-common/blob/b686a5759ed0935da73c58ea8c5a5daaefb10048/include/libretro.h#L151-L181

now your using a touchscreen you would expect a 1:1 match on choords where your touch the screen perhaps change your aspect ratio to suit the pointer setup if that is the issue. As for not touching the screen I have no idea what android is returning
